Batch service 2017-01-01 version release

xingwu1 opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] Validation errors from the [Linter extension for VS Code](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/linter.md) have all been fixed for this spec. (**Note:** for large, previously checked in specs, there will likely be many errors shown. Please contact our team so we can set a timeframe for fixing these errors if your PR is not going to address them).
  - [x] The spec follows the patterns described in the [Swagger good patterns](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-good-patterns.md) document unless the service API makes this impossible.

xingwu1 committed:
Commit the old spec as base.

xingwu1 committed:
New 2017-01-01 version release.

xingwu1 commented:
The new guideline require the operations to have example. However, there are too many operations for our spec. Is it hard requirement to have example for all the operations in order to check in? or we can add them later.

amarzavery commented:
@xingwu1 - I understand your concern. However, we have found lot of issues and some of them P0 when service teams provided us examples. These bugs were not caught by the sdk tests. It would be a one time thing to grandfather old operations. Moving forward, it would only be additive. These examples will also be used by the doc team to highlight them on the docs.microsoft.com website that is generating the REST API docs from swagger specs. This helps customers with a payload that they can copy/paste and use out of the box with very few modifications.

itowlson commented on batch/2017-01-01.4.0/swagger/BatchService.json:
```diff
@@ -0,0 +1,14079 @@
+          "200": {
+            "headers": {
+              "client-request-id": {
+                "description": "The client-request-id provided by the client during the request. This will be returned only if the return-client-request-id parameter was set to t
```
Per CodeFlow and email, please let's not use this text as it confuses the issue about which task it refers to.  Proposed text from Tamra is: "This task will not be scheduled until all tasks that it depends on have completed successfully. If any of those tasks fail and exhaust their retry counts, this task will never be scheduled."

xingwu1 replied:
resolved.

matthchr replied:
We need to describe what `readOnly` and `readWrite` mean here (probably in a description field)

olydis replied:
Can this be named `Job_ListFromSchedule`? I see that this is not the same as `Job_ListJobs` that should clearly be renamed, but maybe the "jobness" of this schedule is also somewhat implicit. Just an idea, you probably have a better idea how this feels like to customers and what they expect ;-)

olydis replied:
seems we hit the GitHub line number bug again... I commented on line 5211, but my comment appears on line 1507 (just as the comment by @matthchr) for me

matthchr replied:
@olydis The name "JobSchedule" in our service is actually the name of an entity. So using that makes sense in this context because it is "List all jobs (an entity) created by the particular job schedule (another entity)."  

I think it would be confusing to say "ListFromSchedule" because while that reads okay, there is no "Schedule" entity in the service, only a "JobSchedule"

xingwu1 replied:
The JobSchedule is a key term in Batch service. It isn't just combining of 'Job' and 'Schedule' words.

olydis replied:
it smelled like that, alright! 👍

xingwu1 committed:
Updated some wording.

olydis commented:
General comment: I see more than 100 linter warnings in BatchService.json - but you have ticked the corresponding check box under *Quality of Swagger*!
Most of these warnings are about "missing descriptions", it seems you use `title` instead - I think this is fine, we might update the linter accordingly. However, you did not comment on that at all (there are also other warnings).

olydis commented:
You use `uuid` format a lot. We gave this recommendation previously (sorry!), however we had a meeting with the Resource Manager team and they suggested that one shall not use Guids unless there is a very good reason (i.e. if you as a C# user would have a clear *benefit* of dealing with a Guid -- a simple string is usually way better). So please remove format `uuid` unless it is absolutely required. /cc @amarzavery

matthchr commented:
@olydis 

For the `title` vs `description` thing, I already filed a bug for the linter [here](https://github.com/Azure/autorest/issues/1767)

For `uuid` -- We are using this for the `client-request-id` and `request-id` fields, which our service expects to be GUIDs. Since you'll get rejected if you come with something that isn't a GUID, it makes sense to use `uuid` here right - As a matter of fact previously we said "This string must be in the form of a GUID" -- which is awkward because if they didn't read the docs they would've missed that. Also note: This swagger specification is for a data plane, _not_ a resource manager spec, so guidance from the ARM team isn't really useful to us.

xingwu1 commented:
@olydis There are only 3 different kind of lint warnings, no error. 

1. As @matthchr pointed out, title vs description warning doesn't make sense. 
2. We already reply to about "JobSchedule" is a key word.
3. This is data plane spec, the flatten rule doesn't make too much sense in our spec.

I didn't comment them since I already clean up all the error/warning make sense to our spec. Generally, I think the lint rules are set too restrict for data plane spec.

olydis commented:
@matthchr 
>This swagger specification is for a data plane, not a resource manager spec, so guidance from the ARM team isn't really useful to us.

I'm aware of that, but at the end of the day consistency is still something to consider.
Also `client-request-id`/`request-id` is something widely used across the services, including for example *arm-search* (and yes, they currently all model it as GUID!).

This is the first PR where I see `client-request-id`/`request-id` with `format: uuid` since I know of the ARM team's decision so I thought we'd revise it now. I'm fine with whatever we decide as long as we enforce it consistently. ;-)

@ravbhatnagar what do you think? See @matthchr's [points](https://github.com/Azure/azure-rest-api-specs/pull/892#issuecomment-276743710) for using GUID.

xingwu1 committed:
Add some examples and improving wording

matthchr commented:
@olydis Flipping it around, what's the advantage of using `string` over `uuid`? It allows no more accepted values in our case (since the service only accepts `uuid`'s).

olydis commented:
I was not part of that discussion, I understand your point. IMO one could also argue that `string` vs `Guid` is a code generation decision - the Swagger should describe the wire and not worry about what kind of clients are generated from it.

xingwu1 commented:
@olydis @amarzavery 
Any further work do you except us to do before you sign off the spec?

olydis commented on batch/2017-01-01.4.0/swagger/BatchService.json:
```diff
@@ -0,0 +1,14081 @@
+          "200": {
+            "headers": {
+              "client-request-id": {
+                "description": "The client-request-id provided by the client during the request. This will be returned only if the return-client-request-id parameter was set to t
```
The enum set was reduced (and the enum renamed), this is a breaking change.
Don't forget to bump up the major version of the SDK.
Also, please model this enum as `modelAsString: true` (it should have been modeled that way all along).

olydis replied:
Same as above. Please also extract the type definition (deduplicate code!) which should simplify both the authoring and the reviewing process.

olydis replied:
@line 13234

olydis replied:
(line 13254)

xingwu1 replied:
@olydis 
This is major version update (from 3.1 to 4.0).
Why we have to set the modelAsString to true? we would like to use it as enum as possible.

xingwu1 replied:
@olydis 
I don't know how we can extract the simple (enum) to the type definition. Could you give me a hint for extracting this enum?

matthchr replied:
A couple of things:
1. Yes, this is a breaking API change. We will be bumping the major version across all of our clients. There are other breaking changes in here other than just this enum.
2. For `modelAsString`, why would we ever want to set it to `true`? Based on the documentation (here)[https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-enum] and from my understanding of what AutoRest does, `modelAsString: true` makes a method that looks like: `MyMethod(string myEnum)` whereas using `modelAsString: false` makes a method that looks like `MyMethod(MyEnum myEnum)`. The second method signature is clearly superior in strongly typed languages such as C# and Java as it limits invalid inputs and catches errors at compile time. Why would we want to give up that validation/logic and use `modelAsString: true`? I know that this is thinking about making an augmented enum type see issue [1591](https://github.com/Azure/autorest/issues/1591) but until that exists across languages we want to use a type which is as idiomatic as possible in the languages our customers are using, and that's an `enum`.

olydis replied:
`definitions` section:
```
"SubtaskState": {
  "type": "string",
  "enum": [
    "preparing",
    "running",
    "completed"
  ],
  "x-ms-enum": {
    "name": "SubtaskState",
    "modelAsString": false
  }
},
```

Usage:
```
"previousState": {
  "title": "The previous state of the subtask.",
  "description": "This property is not set if the subtask is in its initial running state.",
  "$ref": "#/definitions/SubtaskState"
},
```

olydis replied:
This enum (anything that talks about some state) sounds like it could change its set of values (next month the service decides to also have a state "waiting"/"paused" or remove yet another value), so yes, this is all about the enum experience! The worst experience for customers is to have the SDK throw because the server returned an unknown enum value because

a) the server/service was updated (i.e. the enum has grown) or
b) the queried value originated from a time where it was still supported (i.e. the enum has shrunk as in your case)

We deal with these issues a lot. This is all about reducing the "breakingness" of such a tiny change.

xingwu1 replied:
@olydis 
Thanks. Working on it.

matthchr replied:
@olydis Our server guarantees that it will not return unknown Enum values to clients that won't understand them. It does this using the `api-version` provided in the URI. In this particular Enum's case, it actually should never have had `Active` state in it to begin with as the server never returned that value. It was a mistake made when authoring the Swagger specification, so there is no risk to past clients.

You can read my comment [here](https://github.com/Azure/autorest/issues/1591#issuecomment-260433709) for more into what exactly the Batch service does and where it would be useful to use `modelAsString` (answer: It might be useful on enum's which are not "State" enums).

Still, I would be _really_ hesitant to make the switch to `modelAsString` today since the current experience for `modelAsString` is so poor and in the past 2 years we haven't run into any issues with it since the non-state enums in our service are pretty much static. If we do make a change to a non-state enum today, we will work around it on the server using `api-version` to avoid breaking clients while still maintaining a good client experience.

Once you guys have support for [1591](https://github.com/Azure/autorest/issues/1591), which was also asked for by me in [733](https://github.com/Azure/autorest/issues/733) and by @brjohnstmsft in [290](https://github.com/Azure/autorest/issues/290) then we will move to that.

xingwu1 replied:
@olydis 
I just extract the duplicated enum type definition to root. Thank for the hint. It is much better.

xingwu1 replied:
@olydis 
The vanilla swagger spec only treat the enum as string. However, the Autorest has ability to map to real enum type in the language support enum is great user experience improvement. I really don't want to strict the spec to only use string for enum of due to some server devs don't want to update the API version. 
The possible move for AutoRest is set the default value of `modelAsString` to true to keep the default behavior same as vanilla swagger spec.

olydis commented:
Okay, I received clarification regarding the Guid recommendation, it applies to body parameters, your use case is perfectly fine 👍

xingwu1 committed:
Move the duplicate enum types to the root of definitions.

olydis commented on arm-batch/2017-01-01/swagger/BatchManagement.json:
```diff
@@ -0,0 +1,1563 @@
+        "poolAllocationMode": {
+          "type": "string",
+          "title": "The allocation mode to use for creating pools in the Batch account.",
+          "description": "The pool allocation mode also affects how clients may authenticate to the Batch Service API. If the mode is BatchService, clients may authenticate using access keys or Azure Active Directory. If the mode is UserSubscription, clients must use Azure Active Directory. The default is BatchService.",
```
Don't say "the default is ...", specify the `default` using swagger instead.

olydis replied:
oh and reference `PoolAllocationMode` from below

matthchr replied:
Agree that "the default is..." is unfortunate. But sadly `default` in Swagger causes the parameter to be serialized across the wire, which is awkward when you have a ton of parameters with default behavior. See issue [1607](https://github.com/Azure/autorest/issues/1607).

Until that issue is fixed I'm hesitant to sprinkle `default` everywhere. But we will definitely make a note of it and consider fixing it in a future release.

olydis replied:
Ah yes... that one will soon be configurable. Alrighdy

olydis commented on arm-batch/2017-01-01/swagger/BatchManagement.json:
```diff
@@ -0,0 +1,1563 @@
+      ],
+      "description": "Contains information about the auto storage account associated with a Batch account."
+    },
+    "BatchAccountProperties": {
```
shouldn't those other properties be `readOnly` as well?

xingwu1 replied:
Add some properties with readonly flag, but not all the them. The reason is the property can't have both readonly and required flag, so if we use readonly, the simple type will become nullable which is very bad user experience. User have to use account.State.**Value** to get the value. We filed an AutoRest issue about this.

olydis replied:
There is `"x-nullable": false` if it is about the user experience. Since this is a response-only model (correct?), it would be somewhat weird to make some properties readonly while others are not...

xingwu1 replied:
@olydis Thank you for the hint. It works very well. I updated the schedule with correctly readonly flag.

olydis commented:
This is not a must but it looks like you could deduplicate A LOT of Swagger by extracting parameters! E.g. you have tons of identical `client-request-id`, `ocp-date`, `If-Match`, ... parameters. For some you have updated just a detail, and already there are hundreds of changes and a huge diff! See the usage of `"x-ms-parameter-location": "method"` in https://github.com/Azure/azure-rest-api-specs/blob/7171f61ec60c554aa12c557abab4fa33b1b3e827/arm-insights/2016-03-01/swagger/insightsManagementClient_AlertRulesIncidents.json for an example.

matthchr commented:
@olydis Yeah, when we originally wrote the spec, `x-ms-parameter-location` didn't support `method`. We are _definitely_ going to make the change to de-duplicate a ton of stuff. We are also going to move to YAML I think. But since all of that is not going to impact the generated code we will do it a bit later. Right now we are just trying to get this REST API version out because the service will be ready shortly and we want to ship updated clients ASAP. If we redo the whole Swagger specification now it'll take too much time. But rest assured (for our own sanity) we will be doing as you suggest soon.

olydis commented:
Will sync with the review team what our current status of automated checking is (using your provided examples and such), other than that and my minor comments above this looks good!

olydis commented on batch/2017-01-01.4.0/swagger/BatchService.json:
```diff
@@ -0,0 +1,13987 @@
+          "200": {
+            "headers": {
+              "client-request-id": {
+                "description": "The client-request-id provided by the client during the request. This will be returned only if the return-client-request-id parameter was set to t
```
Don't use "Patch" as a method name, the guideline is to use "Update" here (the summary even talks about updating anyway!)

olydis replied:
oh I see you use "POST" to actually create a `JobSchedule`? Hmm...

xingwu1 committed:
Update with feedback.
1. Change the ErrorBody to standard CloudError
2. Move the duplicated poolAllocationMode to common place
3. Add readonly to some properties.

xingwu1 committed:
Add "x-nullable" to enable properties with "readonly".

olydis reviewed (approved):
LGTM 👍 the tools seem to be happy as well (apart form what we've talked about)

xingwu1 commented:
@olydis @amarzavery Could you merge the spec? Our service is releasing the new version.

amarzavery commented:
@xingwu1 - The model validator found issues. Please check [here](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/198993100#L2339)

amarzavery commented on arm-batch/2017-01-01/examples/BatchAccountCreate_Default.json:
```diff
@@ -0,0 +1,39 @@
+{
+  "parameters": {
+    "name": "sampleacct",
```
the parameter name in the spec is: "accountName"

amarzavery commented on arm-batch/2017-01-01/examples/BatchAccountDelete.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "parameters": {
+    "name": "sampleacct",
```
the parameter name in the spec is: "accountName"

amarzavery commented on arm-batch/2017-01-01/examples/BatchAccountGet.json:
```diff
@@ -0,0 +1,30 @@
+{
+  "parameters": {
+    "name": "sampleacct",
```
the parameter name in the spec is: "accountName"

xingwu1 commented:
@amarzavery I fixed the issue.

AutorestCI commented:
No modification for Ruby

AutorestCI commented:
No modification for Python

AutorestCI commented:
No modification for NodeJS
